File: app/src/main/java/org/qp/android/questopiabundle/QuestopiaBundle.java
package org.qp.android.questopiabundle;

import android.app.Service;
import android.content.Intent;
import android.net.Uri;
import android.os.Handler;
import android.os.IBinder;
import android.os.RemoteException;
import android.util.Log;

import androidx.annotation.Nullable;

import com.anggrayudi.storage.file.DocumentFileCompat;

import org.qp.android.questopiabundle.lib.LibGameRequest;
import org.qp.android.questopiabundle.lib.LibNDKProxyImpl;
import org.qp.android.questopiabundle.lib.LibProxyImpl;
import org.qp.android.questopiabundle.lib.LibRefIRequest;
import org.qp.android.questopiabundle.lib.LibTypeDialog;
import org.qp.android.questopiabundle.lib.LibTypeWindow;

public class QuestopiaBundle extends Service implements GameInterface {

    private final Handler counterHandler = new Handler();
    private final Handler counterNDKHandler = new Handler();
    private LibProxyImpl libProxy;
    private LibNDKProxyImpl libNDKProxy;
    private AsyncCallbacks callbacks;
    private volatile int counterInterval = 500;
    private final Runnable counterTask = new Runnable() {
        @Override
        public void run() {
            libProxy.executeCounter();
            counterHandler.postDelayed(this, counterInterval);
        }
    };
    private final Runnable counterNDKTask = new Runnable() {
        @Override
        public void run() {
            libNDKProxy.executeCounter();
            counterNDKHandler.postDelayed(this, counterInterval);
        }
    };
    private int mLibVersion;

    @Override
    public boolean isPlayingFile(String filePath) {
        try {
            if (callbacks == null) return false;
            return callbacks.isPlayingFile(filePath);
        } catch (Exception e) {
            Log.e(this.getClass().getSimpleName(), "Error", e);
            return false;
        }
    }

    @Override
    public void closeAllFiles() {
        try {
            if (callbacks == null) return;
            callbacks.closeAllFiles();
        } catch (Exception e) {
            Log.e(this.getClass().getSimpleName(), "Error", e);
        }
    }

    @Override
    public void closeFile(String filePath) {
        try {
            if (callbacks == null) return;
            callbacks.closeFile(filePath);
        } catch (Exception e) {
            Log.e(this.getClass().getSimpleName(), "Error", e);
        }
    }

    @Override
    public void playFile(String path, int volume) {
        try {
            if (callbacks == null) return;
            callbacks.playFile(path, volume);
        } catch (Exception e) {
            Log.e(this.getClass().getSimpleName(), "Error", e);
        }
    }

    public void setCallback() {
        if (mLibVersion == 570) {
            counterNDKHandler.postDelayed(counterNDKTask, counterInterval);
        } else if (mLibVersion == 592) {
            counterHandler.postDelayed(counterTask, counterInterval);
        }
    }

    public void removeCallback() {
        counterHandler.removeCallbacks(counterTask);
        counterNDKHandler.removeCallbacks(counterNDKTask);
    }

    @Override
    public void doChangeCurrGameDir(Uri newGameDirUri) {
        try {
            if (callbacks == null) return;
            callbacks.sendChangeCurrGameDir(newGameDirUri);
        } catch (Exception e) {
            Log.e(this.getClass().getSimpleName(), "Error", e);
        }
    }

    @Override
    public void doRefresh(LibRefIRequest request) {
        try {
            if (callbacks == null) return;
            if (mLibVersion == 570) {
                callbacks.sendLibRef(new LibResult<>(request));
                callbacks.sendLibGameState(new LibResult<>(libNDKProxy.getGameState()));
            } else if (mLibVersion == 592) {
                callbacks.sendLibRef(new LibResult<>(request));
                callbacks.sendLibGameState(new LibResult<>(libProxy.getGameState()));
            }
        } catch (Exception e) {
            Log.e("QuestopiaBundle", "Error", e);
        }
    }

    @Override
    public LibDialogRetValue showLibDialog(LibTypeDialog dialog, String inputMessage) {
        var wrap = new LibDialogRetValue();

        if (callbacks == null) {
            if (dialog == LibTypeDialog.DIALOG_MENU) {
                wrap.outNumValue = -1;
            } else {
                wrap.outTextValue = "";
            }
            return wrap;
        }

        try {
            return callbacks.doOnShowDialog(new LibResult<>(dialog), inputMessage);
        } catch (RemoteException e) {
            Log.e("QuestopiaBundle", "Error", e);
        }

        if (dialog == LibTypeDialog.DIALOG_MENU) {
            wrap.outNumValue = -1;
        } else {
            wrap.outTextValue = "";
        }
        return wrap;
    }

    @Override
    public void changeVisWindow(LibTypeWindow type, boolean show) {
        if (callbacks == null) return;
        try {
            callbacks.doChangeVisWindow(new LibResult<>(type), show);
        } catch (RemoteException e) {
            Log.e("QuestopiaBundle", "Error", e);
        }
    }

    @Override
    public void setCountInter(int delayMillis) {
        counterInterval = delayMillis;
    }

    @Override
    public void doWithCounterDisabled(Runnable runnable) {
        if (mLibVersion == 570) {
            counterNDKHandler.removeCallbacks(counterNDKTask);
            runnable.run();
            counterNDKHandler.postDelayed(counterNDKTask, counterInterval);
        } else if (mLibVersion == 592) {
            counterHandler.removeCallbacks(counterTask);
            runnable.run();
            counterHandler.postDelayed(counterTask, counterInterval);
        }
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return new IQuestopiaBundle.Stub() {
            @Override
            public String versionPlugin() throws RemoteException {
                return "0.0.1";
            }

            @Override
            public String titlePlugin() throws RemoteException {
                return "Questopia Bundle";
            }

            @Override
            public String authorPlugin() throws RemoteException {
                return "l3ger0j";
            }

            @Override
            public void startNativeLib(int libVer) throws RemoteException {
                mLibVersion = libVer;
                if (mLibVersion == 570) {
                    libNDKProxy = new LibNDKProxyImpl(QuestopiaBundle.this);
                    setCallback();

                    libNDKProxy.setGameInterface(QuestopiaBundle.this);
                    libNDKProxy.startLibThread();
                } else if (mLibVersion == 592) {
                    libProxy = new LibProxyImpl(QuestopiaBundle.this);
                    setCallback();

                    libProxy.setGameInterface(QuestopiaBundle.this);
                    libProxy.startLibThread();
                }
            }

            @Override
            public void stopNativeLib(int libVer) throws RemoteException {
                mLibVersion = libVer;
                if (mLibVersion == 570) {
                    libNDKProxy.setGameInterface(null);
                    libNDKProxy.stopLibThread();
                    libNDKProxy = null;

                    stopSelf();
                } else if (mLibVersion == 592) {
                    libProxy.setGameInterface(null);
                    libProxy.stopLibThread();
                    libProxy = null;

                    stopSelf();
                }
            }

            @Override
            public void runGameIntoLib(long gameId,
                                       String gameTitle,
                                       Uri gameDirUri,
                                       Uri gameFileUri) throws RemoteException {
                Log.i(this.getClass().getSimpleName(), String.valueOf(DocumentFileCompat.getAccessibleAbsolutePaths(getBaseContext())));
                Log.d(this.getClass().getSimpleName(), "Debug: " + "\nGameID|" + gameId + "\nGameTitle|" + gameTitle + "\nGameDirUri|" + gameDirUri + "\nGameFileUri|" + gameFileUri);
                if (mLibVersion == 570) {
                    libNDKProxy.runGame(gameId, gameTitle, gameDirUri, gameFileUri);
                } else if (mLibVersion == 592) {
                    libProxy.runGame(gameId, gameTitle, gameDirUri, gameFileUri);
                }
            }

            @Override
            public void onActionClicked(int index) throws RemoteException {
                if (mLibVersion == 570) {
                    libNDKProxy.onActionClicked(index);
                } else if (mLibVersion == 592) {
                    libProxy.onActionClicked(index);
                }
            }

            @Override
            public void onObjectClicked(int index) throws RemoteException {
                if (mLibVersion == 570) {
                    libNDKProxy.onObjectSelected(index);
                } else if (mLibVersion == 592) {
                    libProxy.onObjectSelected(index);
                }
            }

            @Override
            public void doLibRequest(LibResult gameRequest, String codeToExec, Uri fileUri) throws RemoteException {
                if (mLibVersion == 570) {
                    var libGameReq = (LibGameRequest) gameRequest.value;
                    switch (libGameReq) {
                        case LOAD_FILE -> doWithCounterDisabled(() -> libNDKProxy.loadGameState(fileUri));
                        case SAVE_FILE -> libNDKProxy.saveGameState(fileUri);
                        case USE_EXECUTOR -> libNDKProxy.onUseExecutorString();
                        case USE_INPUT -> libNDKProxy.onInputAreaClicked();
                        case RESTART_GAME -> libNDKProxy.restartGame();
                        case EXECUTE_CODE -> libNDKProxy.execute(codeToExec);
                    }
                } else if (mLibVersion == 592) {
                    var libGameReq = (LibGameRequest) gameRequest.value;
                    switch (libGameReq) {
                        case LOAD_FILE -> doWithCounterDisabled(() -> libProxy.loadGameState(fileUri));
                        case SAVE_FILE -> libProxy.saveGameState(fileUri);
                        case USE_EXECUTOR -> libProxy.onUseExecutorString();
                        case USE_INPUT -> libProxy.onInputAreaClicked();
                        case RESTART_GAME -> libProxy.restartGame();
                        case EXECUTE_CODE -> libProxy.execute(codeToExec);
                    }
                }
            }

            @Override
            public void sendAsync(AsyncCallbacks callbacks) throws RemoteException {
                QuestopiaBundle.this.callbacks = callbacks;
            }
        };
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        libProxy = null;
        libNDKProxy = null;
        removeCallback();
    }
}

File: app/src/main/java/org/qp/android/questopiabundle/QuestopiaBundle.kt
package org.qp.android.questopiabundle

import android.app.Service
import android.content.Intent
import android.net.Uri
import android.os.IBinder
import android.os.RemoteException
import android.util.Log
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import org.qp.android.questopiabundle.dto.LibGameState
import org.qp.android.questopiabundle.lib.LibGameRequest
import org.qp.android.questopiabundle.lib.LibRefIRequest
import org.qp.android.questopiabundle.lib.LibTypeDialog
import org.qp.android.questopiabundle.lib.LibTypeWindow
import org.qp.android.questopiabundle.lib.impl.LibAlphaProxyImpl
import org.qp.android.questopiabundle.lib.impl.LibBravoProxyImpl
import org.qp.android.questopiabundle.lib.impl.LibCharlieProxyImpl
import kotlin.concurrent.Volatile

class QuestopiaBundle : Service(), GameInterface {
    private lateinit var counterMainJob: Job
    private lateinit var counterNDKJob: Job
    private lateinit var counterSNXJob: Job
    private val libHandlerScope = CoroutineScope(Dispatchers.Default)
    private val libAlphaProxy: LibAlphaProxyImpl = LibAlphaProxyImpl(this@QuestopiaBundle, this)
    private val libBravoProxy: LibBravoProxyImpl = LibBravoProxyImpl(this@QuestopiaBundle, this)
    private val libCharlieProxy: LibCharlieProxyImpl = LibCharlieProxyImpl(this@QuestopiaBundle, this)
    private lateinit var callbacks: AsyncCallbacks

    @Volatile private var counterInterval = 500L
    private val counterMainTask: suspend CoroutineScope.() -> Unit = {
        while (isActive) {
            libAlphaProxy.executeCounter()
            delay(counterInterval)
        }
    }
    private val counterNDKTask: suspend CoroutineScope.() -> Unit = {
        while (isActive) {
            libBravoProxy.executeCounter()
            delay(counterInterval)
        }
    }
    private val counterSNXTask: suspend CoroutineScope.() -> Unit = {
        while (isActive) {
            libCharlieProxy.executeCounter()
            delay(counterInterval)
        }
    }

    @Volatile private var mLibVersion = 570

    override fun requestReceiveFile(filePath: String): Uri {
        try {
            return callbacks.requestReceiveFile(filePath)
        } catch (e: Exception) {
            Log.e(javaClass.simpleName, "Error", e)
            return Uri.EMPTY
        }
    }

    override fun requestCreateFile(dirUri: Uri, path: String): Uri {
        try {
            return callbacks.requestCreateFile(dirUri, path) ?: Uri.EMPTY
        } catch (e: Exception) {
            Log.e(javaClass.simpleName, "Error", e)
            return Uri.EMPTY
        }
    }

    override fun isPlayingFile(filePath: String): Boolean {
        try {
            return callbacks.isPlayingFile(filePath)
        } catch (e: Exception) {
            Log.e(javaClass.simpleName, "Error", e)
            return false
        }
    }

    override fun closeAllFiles() {
        try {
            callbacks.closeAllFiles()
        } catch (e: Exception) {
            Log.e(javaClass.simpleName, "Error", e)
        }
    }

    override fun closeFile(filePath: String?) {
        try {
            callbacks.closeFile(filePath)
        } catch (e: Exception) {
            Log.e(javaClass.simpleName, "Error", e)
        }
    }

    override fun playFile(path: String?, volume: Int) {
        try {
            callbacks.playFile(path, volume)
        } catch (e: Exception) {
            Log.e(javaClass.simpleName, "Error", e)
        }
    }

    private fun removeCallback() {
        counterMainJob.cancel()
        counterNDKJob.cancel()
        counterSNXJob.cancel()
    }

    override fun doChangeCurrGameDir(newGameDirUri: Uri?) {
        try {
            callbacks.sendChangeCurrGameDir(newGameDirUri)
        } catch (e: Exception) {
            Log.e(javaClass.simpleName, "Error", e)
        }
    }

    override fun doUpdateState(request: LibRefIRequest) {
        try {
            when (mLibVersion) {
                570 -> {
                    callbacks.updateState(LibResult<LibRefIRequest?>(request), LibResult<LibGameState?>(libBravoProxy.gameState))
                }

                575 -> {
                    callbacks.updateState(LibResult<LibRefIRequest?>(request), LibResult<LibGameState?>(libCharlieProxy.gameState))
                }

                592 -> {
                    callbacks.updateState(LibResult<LibRefIRequest?>(request), LibResult<LibGameState?>(libAlphaProxy.gameState))
                }
            }
        } catch (e: Exception) {
            Log.e("QuestopiaBundle", "Error", e)
        }
    }

    override fun showLibDialog(dialog: LibTypeDialog?, inputString: String?): LibDialogRetValue? {
        try {
            return callbacks.doOnShowDialog(LibResult(dialog), inputString)
        } catch (e: RemoteException) {
            Log.e("QuestopiaBundle", "Error", e)
            return LibDialogRetValue()
        }
    }

    override fun changeVisWindow(type: LibTypeWindow?, show: Boolean) {
        try {
            callbacks.doChangeVisWindow(LibResult<LibTypeWindow?>(type), show)
        } catch (e: RemoteException) {
            Log.e("QuestopiaBundle", "Error", e)
        }
    }

    override fun setCountInter(delayMillis: Int) {
        counterInterval = delayMillis.toLong()
    }

    override fun doWithCounterDisabled(runnable: Runnable?) {
        when (mLibVersion) {
            570 -> {
                counterNDKJob.cancel()
                runnable?.run()
                counterNDKJob = libHandlerScope.launch(block = counterNDKTask)
            }

            575 -> {
                counterSNXJob.cancel()
                runnable?.run()
                counterSNXJob = libHandlerScope.launch(block = counterSNXTask)
            }

            592 -> {
                counterMainJob.cancel()
                runnable?.run()
                counterMainJob = libHandlerScope.launch(block = counterMainTask)
            }
        }
    }

    override fun onBind(intent: Intent): IBinder = object : IQuestopiaBundle.Stub() {
        @Throws(RemoteException::class)
        override fun versionPlugin(): String {
            return BuildConfig.VERSION_NAME
        }

        @Throws(RemoteException::class)
        override fun titlePlugin(): String {
            return "Questopia Bundle"
        }

        @Throws(RemoteException::class)
        override fun authorPlugin(): String {
            return "l3ger0j"
        }

        @Throws(RemoteException::class)
        override fun startNativeLib(libVer: Int) {
            mLibVersion = libVer
            when (mLibVersion) {
                570 -> {
                    counterNDKJob = libHandlerScope.launch(block = counterNDKTask)
                    libBravoProxy.startLibThread()
                }

                575 -> {
                    counterSNXJob = libHandlerScope.launch(block = counterSNXTask)
                    libCharlieProxy.startLibThread()
                }

                592 -> {
                    counterMainJob = libHandlerScope.launch(block = counterMainTask)
                    libAlphaProxy.startLibThread()
                }
            }
        }

        @Throws(RemoteException::class)
        override fun stopNativeLib(libVer: Int) {
            mLibVersion = libVer
            when (mLibVersion) {
                570 -> libBravoProxy.stopLibThread()

                575 -> libCharlieProxy.stopLibThread()

                592 -> libAlphaProxy.stopLibThread()
            }
            removeCallback()
        }

        @Throws(RemoteException::class)
        override fun runGameIntoLib(
            gameId: Long,
            gameTitle: String,
            gameDirUri: Uri,
            gameFileUri: Uri
        ) {
            when (mLibVersion) {
                570 -> libBravoProxy.runGame(gameId, gameTitle, gameDirUri, gameFileUri)

                575 -> libCharlieProxy.runGame(gameId, gameTitle, gameDirUri, gameFileUri)

                592 -> libAlphaProxy.runGame(gameId, gameTitle, gameDirUri, gameFileUri)
            }
        }

        @Throws(RemoteException::class)
        override fun onActionClicked(index: Int) {
            when (mLibVersion) {
                570 -> libBravoProxy.onActionClicked(index)

                575 -> libCharlieProxy.onActionClicked(index)

                592 -> libAlphaProxy.onActionClicked(index)
            }
        }

        @Throws(RemoteException::class)
        override fun onObjectClicked(index: Int) {
            when (mLibVersion) {
                570 -> libBravoProxy.onObjectSelected(index)

                575 -> libCharlieProxy.onObjectSelected(index)

                592 -> libAlphaProxy.onObjectSelected(index)
            }
        }

        @Throws(RemoteException::class)
        override fun doLibRequest(gameRequest: LibResult<*>, codeToExec: String, fileUri: Uri) {
            when (mLibVersion) {
                570 -> {
                    val libGameReq = gameRequest.value as LibGameRequest
                    when (libGameReq) {
                        LibGameRequest.LOAD_FILE -> doWithCounterDisabled {
                            libBravoProxy.loadGameState(fileUri)
                        }

                        LibGameRequest.SAVE_FILE -> libBravoProxy.saveGameState(fileUri)
                        LibGameRequest.USE_EXECUTOR -> libBravoProxy.onUseExecutorString()
                        LibGameRequest.USE_INPUT -> libBravoProxy.onInputAreaClicked()
                        LibGameRequest.RESTART_GAME -> libBravoProxy.restartGame()
                        LibGameRequest.EXECUTE_CODE -> libBravoProxy.execute(codeToExec)
                    }
                }

                575 -> {
                    val libGameReq = gameRequest.value as LibGameRequest
                    when (libGameReq) {
                        LibGameRequest.LOAD_FILE -> doWithCounterDisabled {
                            libCharlieProxy.loadGameState(fileUri)
                        }

                        LibGameRequest.SAVE_FILE -> libCharlieProxy.saveGameState(fileUri)
                        LibGameRequest.USE_EXECUTOR -> libCharlieProxy.onUseExecutorString()
                        LibGameRequest.USE_INPUT -> libCharlieProxy.onInputAreaClicked()
                        LibGameRequest.RESTART_GAME -> libCharlieProxy.restartGame()
                        LibGameRequest.EXECUTE_CODE -> libCharlieProxy.execute(codeToExec)
                    }
                }

                592 -> {
                    val libGameReq = gameRequest.value as LibGameRequest
                    when (libGameReq) {
                        LibGameRequest.LOAD_FILE -> doWithCounterDisabled {
                            libAlphaProxy.loadGameState(fileUri)
                        }

                        LibGameRequest.SAVE_FILE -> libAlphaProxy.saveGameState(fileUri)
                        LibGameRequest.USE_EXECUTOR -> libAlphaProxy.onUseExecutorString()
                        LibGameRequest.USE_INPUT -> libAlphaProxy.onInputAreaClicked()
                        LibGameRequest.RESTART_GAME -> libAlphaProxy.restartGame()
                        LibGameRequest.EXECUTE_CODE -> libAlphaProxy.execute(codeToExec)
                    }
                }
            }
        }

        @Throws(RemoteException::class)
        override fun sendAsync(callbacks: AsyncCallbacks?) {
            this@QuestopiaBundle.callbacks = callbacks ?: return
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        removeCallback()
    }
}
